### ISSUE TITLE
Add commandButton component with actionListener

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#270]
We need to add [commandButton](https://www.primefaces.org/showcase/ui/button/commandButton.xhtml) with `actionListener` and `value `field in property editor.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine command button should be separate component. Button doesn't have actionListener, update etc. It is very different component than simple button.

**Comment by rat-moonshine** [comment]
Have you tested the new implementation? I would suggest if you didn't because actionListener and HTML tags updated upon option check. I think the current options are adequate to serve the purpose. We shall re-check if we needs a separate component anyway after others done review.

**Comment by piotrzarzycki21** [comment]
Nope, because I see in the code that will generate command button. - Yes it will work. However you are mixing two different components inside one. In 2 month may come more requirements but we may be in point where you won't be able to split into two classes those one.

IMO what about commandLink - how are you going to implement those one ? It will be also some option in Button ? 

Please talk with @JoelProminic about that. Even from implementation point of view - there is one more "if" condition in Button instead another class. Better have 20 small classes than 20 ifs in one function.

**Comment by rat-moonshine** [comment]
I agree with you that when there shall be more properties get added to the individual components, then perhaps we may need to decouple these two. At this moment, considering the 1/2 property requirements, I think think should serve the purpose. Yes, we shall need to have different component if the sets grown large. 

I haven't thought about commandLink yet, although this seems doable with Button as well (if we want) considering a few basic properties with it. I see a few button types thankfully, and many of them differ by a or two properties basically. Considering property differences shall be good to manage as a separate component, though; i.e. splitButton.

**Comment by piotrzarzycki21** [comment]
Of course it will work, but in a long term it doesn't look good at all to me, from the implementation point of view and UI.

**Comment by JoelProminic** [comment]
I spent some time thinking about this, but I was not able to come to a conclusive solution about which one is "better".   I think the separate controls are better for the code, but the merged controls are potentially better for the UI.  I think we may need to just pick one implementation for now and refactor if we find that we don't like how it works in the long term.   I wrote up some further thoughts about this below (hopefully this won't just cause more confusion):

Keeping the controls separate would keep the properties simpler for each control, but it may require more duplicated logic.  If we can limit the duplication with inheritance, then the separate controls will probably make for better code.

Having merged controls feels a little more intuitive from the UI perspective.  Instead of having to decide which button does what they want first, the user can add a button, and then they will be prompted for what it should do.

Keeping the controls separate will add more entries to the control list.  Some implications:
- If the controls are sorted alphabetically, then it may not be easy to find the specific button type that you want.  We could mitigate this by sorting or categorizing the controls by "type" (probably defined by manual tags)
- A new user may use the first button they find and be confused about why it does not work.

From a different perspective:  One of the original ideas with the visual editor was that we would make the controls generic so that they could also be used to generate code in a different language (I think we have moved away from this a bit now that we are focused on Visual Editor).  From this perspective, I think both solutions are starting to get a bit specialized to the PrimeFaces implementation.  If we wanted to make this more generic, it might look something like this (this turned out closer to the merged controls, but it could be organized either way):
- Width
- Height
- ID
- Value
- Tooltip
- Enabled
- Action Type (Open Page, Call Action)
- Action Target
- Parameters

**Comment by JoelProminic** [comment]
I reviewed the examples for (button)[https://www.primefaces.org/showcase/ui/button/button.xhtml] and (commandButton)[https://www.primefaces.org/showcase/ui/button/commandButton.xhtml].

For branstorming, I listed the attributes from the examples.  The attributes that are not currently available are marked with a '*':

Shared:
- *id
- value
- title - tooltip
- *icon (We may want to add support for this later)
- enabled

Button:
- *outcome  - name/path of target xhtml file (bookmarkable URL).  We need to add this, or the Button will not do anything interesting
- *<f:param> - parameters to new page.  This may be a more advanced feature

CommandButton:
- actionListener - A ManagedBean action 
- *update - the ID of the component to update after the AJAX action completes
- *ajax - Whether to make the action asyncrhonous (true/false)

LinkButton:  this looks the same as Button to me - I don't really understand why they are separate
- outcome

SplitButton:  This is a much more complicated case, so it will need to be a separate control.  It may be possible to merge this with a "ButtonBar" control if we choose to merge controls.

**Comment by piotrzarzycki21** [comment]
> Having merged controls feels a little more intuitive from the UI perspective. Instead of having to decide which button does what they want first, the user can add a button, and then they will be prompted for what it should do.
> 
> Keeping the controls separate will add more entries to the control list. Some implications:
> 
> If the controls are sorted alphabetically, then it may not be easy to find the specific button type that you want. We could mitigate this by sorting or categorizing the controls by "type" (probably defined by manual tags)
> A new user may use the first button they find and be confused about why it does not work.

To me intuitive interface shouldn't be achieved by path were code is suffer and in a long term we will have classes which are become [Gods](https://en.wikipedia.org/wiki/God_object). 
Above problem shouldn't be resolved from the UI perspective point of view. UI problems should be taken by UI designer and he should organize our controls in a way that user won't be confused and lost. - I was talking about that with Justin some time ago and he was positive to that approach. 
The amount of controls will grow because there are more than one hundred controls in PrimeFaces - to me it's inevitable that some UI designer will have to help us. 

When I'm looking into the list of missing things I see that we could end up in the UI with super long form which is mixed with properties from two different controls. Behind the stage in the code we will have IF-ish labyrinth to handle that. 

The second thing is documentation - It will be also not pleasant, because if you look into current one you will see that on top of the class we are presenting full Visual Editor XML. Now having one Button which describes two controls we will have two long XMLs. Each property will hate to be described in some way to distinguish between Button and CommandButton.

**Comment by piotrzarzycki21** [comment]
This issue will stay open for the next release. I will write more thoughts here.

**Comment by piotrzarzycki21** [comment]
Let's close on that. What about to have an approach with some class called PrimeFacesButton which will store our common properties. Button and CommandButton will extend that one. - Going with that approach we will have clean documentation. - What we will do on UI - that's can be secondary, but I prefer have two different components. Thoughts ?

**Comment by rat-moonshine** [comment]
I'm closing this issue for now. I'm fine with whichever structure that make sense.

**Comment by piotrzarzycki21** [comment]
I'm not unfortunately. 
- We don't have documented that part
- We didn't reach consensus how to resolve all of that to have documentation clean and understandable

I don't think it is ready for resolve state.

**Comment by piotrzarzycki21** [comment]
This issue could be removed from 1.14.0 and could be progressed in the next release.

**Comment by JoelProminic** [comment]
While testing for the 1.14.0 release, I saw that the Command Button is implemented as a checkbox for the Button control:

![image](https://user-images.githubusercontent.com/24532731/43161286-4e31de62-8f4d-11e8-9c70-80337d404d5d.png)

The generated code looks like this:

```
      <p:button style="width:100px;height:30px;" disabled="false" value="Simple Button" title=""/>
      <p:commandButton style="width:100px;height:30px;" disabled="false" value="Command Button" title="" actionListener="testListener"/>
```

So this feature is already part of the upcoming release.  Do we want to reopen this issue for the documentation?

**Comment by piotrzarzycki21** [comment]
It's not only the documentation problem. It is rather problem with putting documentation for that as part of one component. It won't look good, that's why I'm still not happy with solution where everything is in one class. Above conversation have a lot of ideas which we can use to resolve it. 

Yes I wanted to reopen it and have it resolved for the next release.
